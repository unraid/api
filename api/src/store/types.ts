import { type Devices } from '@app/core/types/states/devices';
import { type Networks } from '@app/core/types/states/network';
import { type NfsShares } from '@app/core/types/states/nfs';
import { type Nginx } from '@app/core/types/states/nginx';
import { type SmbShares } from '@app/core/types/states/smb';
import { type Users } from '@app/core/types/states/user';
import { type Var } from '@app/core/types/states/var';
import { MinigraphStatus, type ArrayDisk, type Share } from '@app/graphql/generated/api/types';
import type { RootState } from '@app/store';
import { type DevicesIni } from './state-parsers/devices';
import { type NetworkIni } from './state-parsers/network';
import { type NfsSharesIni } from './state-parsers/nfs';
import { type NginxIni } from './state-parsers/nginx';
import { type SharesIni } from './state-parsers/shares';
import { type SlotsIni } from './state-parsers/slots';
import { type SmbIni } from './state-parsers/smb';
import { type UsersIni } from './state-parsers/users';
import { type VarIni } from './state-parsers/var';
import { type Subscription } from 'zen-observable-ts';


export enum FileLoadStatus {
	UNLOADED = 'UNLOADED',
	LOADING = 'LOADING',
	LOADED = 'LOADED',
	FAILED_LOADING = 'FAILED_LOADING',
}

export enum MemoryCacheStatus {
	UNCACHED = 'UNCACHED',
	CACHED = 'CACHED',
}

export enum StateFileKey {
	var = 'var',
	devs = 'devs',
	network = 'network',
	nginx = 'nginx',
	shares = 'shares',
	disks = 'disks',
	users = 'users',
	sec = 'sec',
	sec_nfs = 'sec_nfs',
}

export interface StateFileToIniParserMap {
	[StateFileKey.var]: (state: VarIni) => Var;
	[StateFileKey.devs]: (state: DevicesIni) => Devices;
	[StateFileKey.network]: (state: NetworkIni) => Networks;
	[StateFileKey.nginx]: (state: NginxIni) => Nginx;
	[StateFileKey.shares]: (state: SharesIni) => Array<Share>;
	[StateFileKey.disks]: (state: SlotsIni) => Array<ArrayDisk>;
	[StateFileKey.users]: (state: UsersIni) => Users;
	[StateFileKey.sec]: (state: SmbIni) => SmbShares;
	[StateFileKey.sec_nfs]: (state: NfsSharesIni) => NfsShares;
}

export enum DaemonConnectionStatus {
	CONNECTED = 'CONNECTED',
	DISCONNECTED = 'DISCONNECTED',
}

export type StoreSubscriptionHandler = (lastState: RootState | null) => Promise<void>;

export enum CacheKeys {
	checkCloud = 'check-cloud',
	checkDns = 'check-dns',
}

export type DNSCheck = {
	cloudIp: string;
	error: null;
	ttl?: number;
} | { error: Error; cloudIp: null };

export const MOTHERSHIP_CRITICAL_STATUSES: Array<MinigraphStatus> = [
	MinigraphStatus.ERROR_RETRYING,
	MinigraphStatus.PING_FAILURE, 
	MinigraphStatus.PRE_INIT
]

export interface SubscriptionWithSha256 {
	sha256: string;
    subscription: Subscription;
}
export interface SubscriptionWithLastPing extends SubscriptionWithSha256 {
    lastPing: number;
}