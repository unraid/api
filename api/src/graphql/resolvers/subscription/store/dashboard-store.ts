import type { Dashboard } from '@app/common/dashboard/generate-data';
import { DashboardPublisher } from '@app/graphql/resolvers/subscription/jobs/dashboard-jobs';

type DashboardCronJobs = ReturnType<typeof DashboardPublisher.init<DashboardPublisher>>;

export const dashboardStore: {
	lastDataPacketTimestamp: number | null;
	lastDataPacket: Dashboard | null;
	lastDataPacketString: string | null;
	connectedToDashboard: number;
	cronJobs: DashboardCronJobs | null;
} = {
	lastDataPacketTimestamp: null,
	lastDataPacket: null,
	lastDataPacketString: null,
	connectedToDashboard: 0,
	cronJobs: null,
};
