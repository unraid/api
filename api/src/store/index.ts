import { configureStore } from '@reduxjs/toolkit';

import { listenerMiddleware } from '@app/store/listeners/listener-middleware';
import { cache } from '@app/store/modules/cache';
import { configReducer } from '@app/store/modules/config';
import { docker } from '@app/store/modules/docker';
import { dynamicRemoteAccessReducer } from '@app/store/modules/dynamic-remote-access';
import { dynamix } from '@app/store/modules/dynamix';
import { emhttp } from '@app/store/modules/emhttp';
import { mothership } from '@app/store/modules/minigraph';
import { notificationReducer } from '@app/store/modules/notifications';
import { paths } from '@app/store/modules/paths';
import { registration } from '@app/store/modules/registration';
import { remoteGraphQLReducer } from '@app/store/modules/remote-graphql';
import { upnp } from '@app/store/modules/upnp';

export const store = configureStore({
    reducer: {
        config: configReducer,
        dynamicRemoteAccess: dynamicRemoteAccessReducer,
        minigraph: mothership.reducer,
        paths: paths.reducer,
        emhttp: emhttp.reducer,
        registration: registration.reducer,
        remoteGraphQL: remoteGraphQLReducer,
        notifications: notificationReducer,
        cache: cache.reducer,
        docker: docker.reducer,
        upnp: upnp.reducer,
        dynamix: dynamix.reducer,
    },
    middleware: (getDefaultMiddleware) =>
        getDefaultMiddleware({
            serializableCheck: false,
        }).prepend(listenerMiddleware.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const getters = {
    cache: () => store.getState().cache,
    config: () => store.getState().config,
    docker: () => store.getState().docker,
    dynamicRemoteAccess: () => store.getState().dynamicRemoteAccess,
    dynamix: () => store.getState().dynamix,
    emhttp: () => store.getState().emhttp,
    minigraph: () => store.getState().minigraph,
    notifications: () => store.getState().notifications,
    paths: () => store.getState().paths,
    registration: () => store.getState().registration,
    remoteGraphQL: () => store.getState().remoteGraphQL,
    upnp: () => store.getState().upnp,
};
