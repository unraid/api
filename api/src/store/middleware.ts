import { type CurriedGetDefaultMiddleware } from '@reduxjs/toolkit/dist/getDefaultMiddleware';

export const defaultAppMiddleware = (getDefaultMiddleware: CurriedGetDefaultMiddleware) =>
    getDefaultMiddleware({
        serializableCheck: {
            ignoredPaths: ['minigraph.client', 'minigraph.subscriptions', 'dashboard.publishJob'],
            ignoredActions: [
                'minigraph/addSubscription',
                'minigraph/createNewClient/fulfilled',
                'minigraph/setClient',
            ],
        },
    });
