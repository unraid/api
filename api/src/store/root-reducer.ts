import { combineReducers, UnknownAction } from '@reduxjs/toolkit';

import { resetStore } from '@app/store/actions/reset-store.js';
import { cache } from '@app/store/modules/cache.js';
import { configReducer } from '@app/store/modules/config.js';
import { dynamicRemoteAccessReducer } from '@app/store/modules/dynamic-remote-access.js';
import { dynamix } from '@app/store/modules/dynamix.js';
import { emhttp } from '@app/store/modules/emhttp.js';
import { mothershipReducer } from '@app/store/modules/minigraph.js';
import { paths } from '@app/store/modules/paths.js';
import { registrationReducer } from '@app/store/modules/registration.js';
import { remoteGraphQLReducer } from '@app/store/modules/remote-graphql.js';
import { upnp } from '@app/store/modules/upnp.js';

/**
 * Root reducer that combines all slice reducers and handles the reset action.
 * When the reset action is dispatched, all slices will be reset to their initial state.
 */
const appReducer = combineReducers({
    config: configReducer,
    dynamicRemoteAccess: dynamicRemoteAccessReducer,
    minigraph: mothershipReducer,
    paths: paths.reducer,
    emhttp: emhttp.reducer,
    registration: registrationReducer,
    remoteGraphQL: remoteGraphQLReducer,
    cache: cache.reducer,
    upnp: upnp.reducer,
    dynamix: dynamix.reducer,
});

// Define the return type of the combined reducer
type AppState = ReturnType<typeof appReducer>;

export const rootReducer = (state: AppState | undefined, action: UnknownAction): AppState => {
    // When the reset action is dispatched, return undefined to reset all reducers
    if (action.type === resetStore.type) {
        return appReducer(undefined, action);
    }

    // Otherwise, use the combined reducer
    return appReducer(state, action);
};
