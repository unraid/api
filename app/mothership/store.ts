import type WebSocketAsPromised from 'websocket-as-promised';

export const store: {
	relay: (WebSocketAsPromised & { _ws?: WebSocket }) | undefined;
	timeout: number | undefined;
} = {
	relay: undefined,
	timeout: undefined
};
