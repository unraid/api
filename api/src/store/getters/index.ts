import { getters, store } from '@app/store';
import { CacheKeys, type DNSCheck } from '@app/store/types';
import { type CloudResponse } from '@app/graphql/generated/api/types';

export const getCloudCache = (): CloudResponse | undefined => {
	const { nodeCache } = getters.cache();
	return nodeCache.get(CacheKeys.checkCloud);
};

export const getDnsCache = (): DNSCheck | undefined => {
	const { nodeCache } = getters.cache();
	return nodeCache.get(CacheKeys.checkDns);
};

export const hasRemoteSubscription = (sha256: string, state = store.getState()): boolean => {
	return state.remoteGraphQL.subscriptions.some(sub => sub.sha256 === sha256);
}