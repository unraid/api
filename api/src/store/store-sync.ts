import { writeFileSync } from 'fs';
import { join } from 'path';

import { isEqual } from 'lodash-es';

import type { RootState } from '@app/store';
import { NODE_ENV } from '@app/environment';
import { store } from '@app/store';
import { syncInfoApps } from '@app/store/sync/info-apps-sync';
import { syncRegistration } from '@app/store/sync/registration-sync';
import { FileLoadStatus } from '@app/store/types';
import { setupConfigPathWatch } from '@app/store/watch/config-watch';

export const startStoreSync = async () => {
    // The last state is stored so we don't end up in a loop of writing -> reading -> writing
    let lastState: RootState | null = null;

    // Update cfg when store changes
    store.subscribe(async () => {
        const state = store.getState();
        // Config dependent options, wait until config loads to execute
        if (state.config.status === FileLoadStatus.LOADED) {
            // Update registration
            await syncRegistration(lastState);

            // Update docker app counts
            await syncInfoApps(lastState);
        }

        if (
            NODE_ENV === 'development' &&
            !isEqual(state, lastState) &&
            state.paths['myservers-config-states']
        ) {
            writeFileSync(join(state.paths.states, 'config.log'), JSON.stringify(state.config, null, 2));
            writeFileSync(
                join(state.paths.states, 'dynamicRemoteAccess.log'),
                JSON.stringify(state.dynamicRemoteAccess, null, 2)
            );
            writeFileSync(
                join(state.paths.states, 'graphql.log'),
                JSON.stringify(state.minigraph, null, 2)
            );
            writeFileSync(
                join(state.paths.states, 'notifications.log'),
                JSON.stringify(state.notifications, null, 2)
            );
        }

        lastState = state;
    });

    setupConfigPathWatch();
};
