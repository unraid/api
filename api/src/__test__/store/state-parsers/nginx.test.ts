import { join } from 'path';

import { expect, test } from 'vitest';

import type { NginxIni } from '@app/store/state-parsers/nginx';
import { store } from '@app/store';

test('Returns parsed state file', async () => {
    const { parse } = await import('@app/store/state-parsers/nginx');
    const { parseConfig } = await import('@app/core/utils/misc/parse-config');
    const { paths } = store.getState();
    const filePath = join(paths.states, 'nginx.ini');
    const stateFile = parseConfig<NginxIni>({
        filePath,
        type: 'ini',
    });
    expect(parse(stateFile)).toMatchSnapshot();
});
