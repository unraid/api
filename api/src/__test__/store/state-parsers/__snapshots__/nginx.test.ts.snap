// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Returns parsed state file 1`] = `
{
  "certificateName": "*.thisisfourtyrandomcharacters012345678900.myunraid.net",
  "certificatePath": "/boot/config/ssl/certs/certificate_bundle.pem",
  "defaultUrl": "https://Tower.local:4443",
  "fqdnUrls": [
    {
      "fqdn": "192-168-1-150.thisisfourtyrandomcharacters012345678900.myunraid.net",
      "id": null,
      "interface": "LAN",
      "isIpv6": false,
    },
    {
      "fqdn": "85-121-123-122.thisisfourtyrandomcharacters012345678900.myunraid.net",
      "id": null,
      "interface": "WAN",
      "isIpv6": false,
    },
    {
      "fqdn": "10-252-0-1.hash.myunraid.net",
      "id": 0,
      "interface": "WG",
      "isIpv6": false,
    },
    {
      "fqdn": "10-252-1-1.hash.myunraid.net",
      "id": 1,
      "interface": "WG",
      "isIpv6": false,
    },
    {
      "fqdn": "10-253-3-1.hash.myunraid.net",
      "id": 2,
      "interface": "WG",
      "isIpv6": false,
    },
    {
      "fqdn": "10-253-4-1.hash.myunraid.net",
      "id": 3,
      "interface": "WG",
      "isIpv6": false,
    },
    {
      "fqdn": "10-253-5-1.hash.myunraid.net",
      "id": 4,
      "interface": "WG",
      "isIpv6": false,
    },
    {
      "fqdn": "10-100-0-1.hash.myunraid.net",
      "id": 0,
      "interface": "TAILSCALE",
      "isIpv6": false,
    },
    {
      "fqdn": "10-100-0-2.hash.myunraid.net",
      "id": 1,
      "interface": "TAILSCALE",
      "isIpv6": false,
    },
    {
      "fqdn": "10-123-1-2.hash.myunraid.net",
      "id": 0,
      "interface": "CUSTOM",
      "isIpv6": false,
    },
    {
      "fqdn": "221-123-121-112.hash.myunraid.net",
      "id": 1,
      "interface": "CUSTOM",
      "isIpv6": true,
    },
  ],
  "httpPort": 8080,
  "httpsPort": 4443,
  "lanIp": "192.168.1.150",
  "lanIp6": "",
  "lanMdns": "Tower.local",
  "lanName": "Tower",
  "sslEnabled": true,
  "sslMode": "yes",
  "wanAccessEnabled": false,
  "wanIp": "",
}
`;
